util/readpass_file.c

```diff
@@ -0,0 +1,106 @@
+#include <sys/stat.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "insecure_memzero.h"
+#include "warnp.h"
+
+#include "readpass.h"
+
+/* Maximum file length. */
+#define MAXPASSLEN 2048
+
+/**
+ * readpass_file(passwd, filename):
+ * Read a passphrase from ${filename} and return it as a malloced
+ * NUL-terminated string via ${passwd}.  Produce an error if the file is 2048
+ * characters or more, or if it contains any newline \n or \r\n characters
+ * other than at the end of the file.  Do not include the \n or \r\n
+ * characters in the passphrase.
+ */
+int
+readpass_file(char ** passwd, const char * filename)
+{
+	struct stat sb;
+	FILE * fp;
+	char passbuf[MAXPASSLEN];
+	size_t len;
+	size_t eol_pos;
+
+	/* Stat the file. */
+	if (stat(filename, &sb)) {
```

Comment:
There's a race condition here if the file is replaced between the stat call and the fopen.  I would fopen first and then use fstat(fileno(f)).

Reply:
... or for that matter just fopen and fgets, without needing to stat the file at all.